### PR TITLE
BIP141: Add 520 bytes witness stack limit

### PR DESCRIPTION
**Pull request by jl2012**
@sipa

### CODE REVIEW

**Comment by luke-jr** [comment]
??? Why a limit?

**Comment by sipa** [comment]
Why a limit? Because DoS potential. I don't like things that allow
arbitrary size data in consensus code.

Script already can't deal with stack items larger than 73 bytes anyway, and
I don't think we should support more, but with 520 bytes we're certain that
it doesn't remove any former functionality.

**Comment by luke-jr** [comment]
The script itself should be longer than 520 bytes or we get the P2SH regression again...

**Comment by sipa** [comment]
The restriction does not apply to the script itself, only to the input
stack fed to it.

**Comment by jl2012** [comment]
@luke-jr @sipa The script limit for P2WSH is 10000 bytes, right?

**Comment by luke-jr** [comment]
@sipa This is unclear, as the witnessScript is popped off the witness stack...

File: bip-0141.mediawiki [hunk 1]
```diff
@@ -85,20 +85,20 @@ There are two cases in which witness validation logic are triggered. Each case d
...
 * The HASH160 of the public key must match the 20-byte witness program.
 * After normal script evaluation, the signature is verified against the public key with CHECKSIG operation. The verification must result in a single TRUE on the stack.
 
 If the version byte is 0, and the witness program is 32 bytes:
 * It is interpreted as a pay-to-witness-script-hash (P2WSH) program.
 * The witness must consist of an input stack to feed to the script, followed by a serialized script ("witnessScript").
-* The witnessScript is popped off the initial witness stack. SHA256 of the witnessScript must match the 32-byte witness program.
-* The witnessScript is deserialized, and executed after normal script evaluation with the remaining witness stack.
+* The witnessScript (≤ 10,000 bytes) is popped off the initial witness stack. SHA256 of the witnessScript must match the 32-byte witness program.
+* The witnessScript is deserialized, and executed after normal script evaluation with the remaining witness stack (≤ 520 bytes for each stack item).
```
Comment (jl2012) [inline bip-0141.mediawiki]: @luke-jr re: witnessScript size and witness stack size

File: bip-0141.mediawiki [hunk 2]
```diff
@@ -85,20 +85,20 @@ There are two cases in which witness validation logic are triggered. Each case d
...
 * The script must not fail, and result in exactly a single TRUE on the stack.
 
 If the version byte is 0, but the witness program is neither 20 nor 32 bytes, the script must fail.
 
-If the version byte is 1 to 16, no further interpretation of the witness program or witness happens.
+If the version byte is 1 to 16, no further interpretation of the witness program or witness happens, and there is no size restriction for the witness. These versions are reserved for future extensions.
```
Comment (luke-jr) [inline bip-0141.mediawiki]: This seems to leave a valid infinite-sized data?

Comment (sipa) [inline bip-0141.mediawiki]: That's still subject to normal Script validation rules.

Comment (luke-jr) [inline bip-0141.mediawiki]: I do not understand your response.

Comment (sipa) [inline bip-0141.mediawiki]: This text describes data that goes into the scriptPubKey or the P2SH
redeemscript.

It gets executed as normal when spent (with an empty or redeemscript-only
scriptSig), before witness execution takes place. During that first
execution, it is subject to the normal script evaluation rules, which for
example limit the size of pushes. Thus, it is not unlimited.

Comment (luke-jr) [inline bip-0141.mediawiki]: Aha, got it.

Comment (jl2012) [inline bip-0141.mediawiki]: @sipa, I am talking about the witness stack here, not scriptPubKey nor scriptSig. There is no size limit for the witness stack if the witness program is v1 or above

Comment (jl2012) [inline bip-0141.mediawiki]: @luke-jr yes, I think leave a valid infinite-sized data, but v1-v16 witness program is non-standard, so only miners can do it.

Comment (luke-jr) [inline bip-0141.mediawiki]: Hmm, I suppose as long as the costlimit is in place that isn't too bad.

**Comment by jl2012** [comment]
Testing with witnessScript of 10000 and 10001 bytes:
https://www.irccloud.com/pastebin/xAEd96Yk/
https://www.irccloud.com/pastebin/k6yfZjTa/

File: bip-0141.mediawiki (hunk 1 above)
Comment (NicolasDorier) [inline bip-0141.mediawiki]: I would say "witness redeem" instead of witnessScript.
WitnessScript are the pushes for a particular input.

**Comment by andre-amorim** [comment]
@jl2012 From left to right top to bottom coordinates of meaning.. is the first link to 10000 and the 10001 to 2nd link ?

**Comment by jl2012** [comment]
@andre-amorim yes, you can find it by clicking the links

**Comment by andre-amorim** [comment]
Thanks @jl2012 it's 212k on my lp0, yep I printed it out.

**Pull request merged**